### PR TITLE
feat(products): implement user-based like/unlike system (closes #11)

### LINKED ISSUE #11: BACKEND: Implement API support for liking and unliking items
### THIS IS THE BACKEND ISSUE TO 314
Find the sibling Frontend issue here: https://github.com/Cherry-CIC/MVP/issues/314

•	Supports storing which users have liked which items
•	Enables frontend to fetch like counts and user-specific like state

Why
As a user
I want my likes to be stored and associated with my account
So that I can see which items I’ve liked and keep track of them across devices

⸻

Acceptance Criteria
	•	Given a logged-in user
When they like an item
Then their user ID should be associated with that item and reflected in the like count
	•	Given a user has liked an item
When they unlike it
Then the user ID should be removed from the item’s like list and the like count decremented
	•	Given an item is fetched for display
When the requesting user is logged in
Then the response should indicate whether the user has liked it (isLikedByUser = true/false)
	•	Given a user fetches their profile or “liked items” view
When the API is called
Then all items the user has liked should be returned
	•	Given a non-authenticated request to like/unlike
When the API is hit
Then a 401 unauthorised error should be returned

⸻

Proposed Implementation
	•	Data model changes:
	•	Add a likedBy array (or set) to the Item model to store user IDs
	•	Optionally maintain a likeCount field for performance
	•	Endpoints:
	•	POST /items/:id/like — Auth required. Adds user ID to item’s likes.
	•	DELETE /items/:id/like — Auth required. Removes user ID from item’s likes.
	•	GET /user/liked-items — Returns all items liked by the current user (optional for MVP)
	•	Update GET /items/:id (and item listing endpoint) to return isLikedByUser for authenticated users
	•	Security:
	•	Require valid JWT auth (Firebase or other)
	•	Prevent duplicate likes (idempotent POST)
	•	Ensure unlikes on non-liked items don’t error
	•	Considerations:
	•	For MVP, keep it simple: no pagination or like history needed
	•	In future: consider moving likes to a join table (UserLikes) if performance or querying becomes complex

### PR DESCRIPTION
**Pull request by 1337Xcode**
Implement the like/unlike feature for products with full authentication, atomic operations, and backwards compatibility. This addresses the issue #11 and provides the backend API support for the frontend issue Cherry-CIC/MVP#314 (which I suppose isn't implemented yet)

Changes:
- Add POST /api/products/:id/like endpoint (idempotent, authenticated)
- Add DELETE /api/products/:id/like endpoint (idempotent, authenticated)
- Add GET /api/products/user/liked endpoint for fetching user's liked products
- Enhance all GET endpoints to include isLikedByUser and likeCount fields
- Implement user-based tracking with likedBy array in Product model
- Use Firestore atomic operations (arrayUnion/arrayRemove) for race condition prevention
- Add comprehensive test suite with 15+ test cases covering authentication, idempotency, and edge cases
- Update Swagger/OpenAPI documentation for all like-related endpoints

Implementation:
- Product model: Added optional likedBy string array field (backwards compatible)
- Repository layer: Implement addLike(), removeLike(), getProductsLikedByUser() with atomic Firestore operations
- Service layer: Add likeProduct(), unlikeProduct(), getProductsLikedByUser() business logic
- Controller layer: Replace likeProduct(), add unlikeProduct() and getLikedProducts() handlers
- Routes: Configure new endpoints with authentication middleware

Security:
- Authentication required for all like/unlike operations (401 on unauthorized)
- User ID extracted from Firebase auth token (not request body)
- Atomic Firestore operations prevent race conditions
- Idempotent operations (safe to retry)

Acceptance Criteria Met:
- Logged-in users can like/unlike items with user ID association
- Like counts increment/decrement atomically
- Item fetch responses include isLikedByUser boolean for authenticated users
- GET /user/liked returns all items liked by current user
- Unauthenticated requests return 401 unauthorized
- Operations are idempotent (duplicate likes/unlikes handled gracefully)

Backwards Compatibility:
- No breaking changes to existing API endpoints
- Optional likedBy field allows existing products to work without migration
- Preserved existing adjustLikes() method for compatibility
- All existing GET endpoint responses maintained with added fields

**PS:** Since the frontend implementation of this feature has not yet been completed, I figured it was important to ensure that the changes remain backwards compatible :)

Documentation:
- Implementation guide: docs/LIKE_FEATURE_IMPLEMENTATION.md

Refs: #11, Cherry-CIC/MVP#314

### CODE REVIEW

**Comment by CherryCIC** [comment]
Hey @1337Xcode thank you for this! 

@funmi-cod could you take a look at this to help double check this is definitely safe to merge?

**Comment by CherryCIC** [comment]
**Summary**

- Purpose: Add user-based like/unlike support for products (POST /api/products/:id/like, DELETE /api/products/:id/like, GET /api/products/user/liked). Track users who liked a product with a likedBy array on Product documents, expose likeCount and isLikedByUser to callers, and use Firestore atomic ops (arrayUnion / arrayRemove / increment).
- Scope: Model change (likedBy?: string[]), repository methods addLike/removeLike/getProductsLikedByUser, new service methods, controller handlers for like/unlike and for getting liked products, route + swagger updates, tests for the new behavior, documentation (docs/LIKE_FEATURE_IMPLEMENTATION.md).
- Acceptance coverage: Unit/route tests added for a number of happy and edge cases (authentication, idempotency, missing product). Swagger docs and an implementation doc are included.

**High-level assessment — good**

- Authentication is required for like/unlike/get-liked endpoints (controller checks user and routes use authMiddleware).
- Tests cover many typical use-cases and edge-cases (idempotency, missing product, backwards compatibility for products without likedBy).
- The implementation is thoughtfully backwards-compatible (likedBy optional, controllers tolerate missing field).
- Documentation (README-like doc + Swagger) included.

**Main risks, security concerns and missing validations (priority order)**

1) Incorrect idempotency / over-count on repeated likes (HIGH)

- Problem: ProductRepository.addLike always issues update({ likedBy: arrayUnion(userId), likes: increment(1) }) with no check whether userId was already present.
- Consequence: If the same user calls the endpoint twice, arrayUnion prevents duplicate userId in likedBy but increment(1) will run both times -> likes will be incremented twice. That breaks idempotency and will desynchronize likes count from likedBy length.
- Repro steps: call POST /:id/like twice quickly from same user (or have retry on network failure) — likedBy will contain a single entry but likes will reflect both increments.
- Fix: Use a transaction that reads the product and checks membership. Only call increment(1) when userId is not already in likedBy. Example approach:
  - run a transaction: read doc; if !doc.exists -> return null; const likedBy = doc.data().likedBy || []; if (!likedBy.includes(userId)) -> update likedBy with arrayUnion and increment(1); else no-op (or still ensure likedBy present). Return the doc after transaction commit.
  - Alternatively, keep arrayUnion/increment but perform a pre-read and then conditional update inside a transaction — don't issue unconditional increment.

2) Potential race conditions / correctness (MEDIUM)

- arrayUnion + increment in separate update is atomic for that single update, but concurrent clients can both read "user not present" and then both execute increment(1) in their own updates, leading to double increment even if arrayUnion dedups. Transactions help mitigate this (read+conditional write inside single transaction).
- removeLike attempts to read the doc and conditionally decrement only if likedBy.includes(userId) — that's better, but it still does a read and then an update without an explicit transaction; two concurrent unlikes or a concurrent like/unlike could race. Use a transaction to ensure correctness.

3) Leakage of likedBy (privacy) in API responses (HIGH)

- In controller code the product object is spread: {...product, likeCount: ..., isLikedByUser: ...}. That means likedBy remains present in the returned object unless the ResponseHandler explicitly strips it. Exposing likedBy returns internal user IDs to any caller (including other authenticated users) and is a privacy/security leak.
- Fix: Explicitly omit likedBy from public responses. Return only likeCount and isLikedByUser (and other allowed product fields). Or map product to pick allowed keys before sending.

4) Firestore document growth / scalability concern (MEDIUM → HIGH if high-scale)

- Storing an ever-growing likedBy array on product documents can eventually hit Firestore document size limits (1 MiB) and degrade performance for writes/reads on hot documents.
- Also array-contains queries on likedBy to support getProductsLikedByUser can be inefficient at scale.
- Recommendation: For large scale, consider a separate collection per-product likes (subcollection) or a top-level 'likes' collection keyed by (productId, userId) and maintain a likeCount on the product doc. That avoids unbounded arrays and supports efficient fan-out/pagination.

5) Missing / insufficient validations (LOW→MEDIUM)

- Controller currently checks for authenticated user, but:
  - No explicit validation that req.params.id conforms to expected product-id format inside controller — routes use validateProductId, so this may be OK, but ensure validateProductId is always applied to these routes and not bypassable.
  - No checks to protect against extremely large likedBy arrays returned from getProductsLikedByUser (pagination?). GET /api/products/user/liked returns all liked products without pagination — if a user has thousands of likes this can be expensive. Consider adding pagination or limits.
  - No rate limiting / abuse protection for like/unlike endpoints. A malicious client could spam like/unlike causing many writes and charges.
  - No server-side caps preventing likes from dropping below zero when inconsistency occurs (though removeLike tries to avoid decrementing if user not present).
- Tests currently mock the repository and will not detect the addLike increment bug — add integration test against Firestore (or simulate the conditional behavior) to assert idempotency.

6) Error-handling & observability (LOW)

- The code returns helpful messages but consider logging failed update details for investigating inconsistencies (e.g., double count events), and monitoring metrics for likes per second, error rate, and unusual growth of likedBy arrays.

File-by-file feedback and actionable suggestions

``typescript name=docs/LIKE_FEATURE_IMPLEMENTATION.md url=https://github.com/Cherry-CIC/cherry-Backend/blob/main/docs/LIKE_FEATURE_IMPLEMENTATION.md
```
- Doc is thorough and helpful as a reference for frontend/backends. Good to have the recommended Firestore index note.
- Suggest adding a "Migration / data-cleanup" section explaining how to fix inconsistent like counts if present (script to reconcile likes with likedBy length).
- Add a note on scaling considerations and recommended alternative schema for large scale (likes collection or per-user list) and mention privacy considerations (don't return likedBy in public API).

``typescript name=src/modules/products/controllers/productController.ts url=https://github.com/Cherry-CIC/cherry-Backend/blob/main/src/modules/products/controllers/productController.ts
```
- Positive: Consolidates adding likeCount/isLikedByUser into GET endpoints and enforces auth for like/unlike/get-liked.
- Problems & recommendations:
  - Do NOT include likedBy in responses. The controller is spreading the whole product object and adding fields; explicitly strip likedBy before sending. Example:
    - const { likedBy, ...publicProduct } = product; then send { ...publicProduct, likeCount, isLikedByUser }.
  - Reduce repeated code: multiple endpoints compute likeCount/isLikedByUser the same way — extract a small helper (e.g., attachLikeMeta(product, userUid)) to avoid duplication and mistakes.
  - The likeProduct controller now always responds isLikedByUser: true (good), but it trusts the repo's returned product.likes. After fixing repo bugs, ensure response is correct.
  - Validate req.params.id presence and format defensive check here as well (even if route validator exists) — fail-safe programming.

``typescript name=src/modules/products/model/Product.ts url=https://github.com/Cherry-CIC/cherry-Backend/blob/main/src/modules/products/model/Product.ts
```
- likedBy?: string[] added — keep it optional. Document what format user IDs have (Firebase UID).
- Suggest adding a field-level comment that likedBy is private and should never be sent in API responses.

``typescript name=src/modules/products/repositories/ProductRepository.ts url=https://github.com/Cherry-CIC/cherry-Backend/blob/main/src/modules/products/repositories/ProductRepository.ts
```
- Critical issue (see above): addLike unconditionally increments likes while using arrayUnion for likedBy. That breaks idempotency.
- Suggested fixes:
  - Implement addLike using a transaction that reads current likedBy and only increments if userId not present:
    - return await db.runTransaction(async (tx) => { const snap = await tx.get(docRef); if (!snap.exists) return null; const likedBy = snap.data().likedBy || []; if (!likedBy.includes(userId)) { tx.update(docRef, { likedBy: FieldValue.arrayUnion(userId), likes: FieldValue.increment(1), updatedAt: FieldValue.serverTimestamp() }); } else { tx.update(docRef, { updatedAt: FieldValue.serverTimestamp() }); } return ...; });
  - removeLike should also be transactional. Even though it conditionally decrements, simultaneous operations can create races; run in a transaction to be safe.
- getProductsLikedByUser uses where('likedBy', 'array-contains', userId) — works but flag scalability concern in code/comments.
- Add guard to ensure likes never becomes negative (defensive): after decrement, ensure min 0.
- Add unit/integration tests that simulate two concurrent addLike calls to verify transaction prevents double counting (or at least confirm behavior).

``
typescript name=src/modules/products/routes/productRoutes.ts url=https://github.com/Cherry-CIC/cherry-Backend/blob/main/src/modules/products/routes/productRoutes.ts
```
- Good: routes updated and Swagger updated to remove previous boolean payload (the controller now uses separate delete route).
- Confirm authMiddleware is correctly applied and always populates req.user with uid after token verification.
- Swagger: ensure components/securitySchemes contains bearerAuth and that swagger generation includes that global config.
- Consider marking GET /user/liked as paginated, or add query params limit/offset to prevent huge responses.

``typescript name=src/modules/products/services/ProductService.ts url=https://github.com/Cherry-CIC/cherry-Backend/blob/main/src/modules/products/services/ProductService.ts
```
- Thin service layer — ok.
- Ensure service methods include argument validation (non-empty productId string and userId). Add small guards and throw consistent errors so controllers can produce correct HTTP codes.
- Consider adding a reconcile method to recalculate likes based on likedBy array for recovery scripts or admin tasks.

``typescript name=src/modules/products/tests/likeFeature.test.ts url=https://github.com/Cherry-CIC/cherry-Backend/blob/main/src/modules/products/tests/likeFeature.test.ts
```
- Tests are comprehensive at the route/controller layer with mocked repository and firebase auth — good coverage for expected behavior.
- Missing tests that would catch the unconditional increment bug: add a test that simulates two identical POSTs (or a retry) and asserts likes didn't increment twice. That requires mocking repository behavior to emulate actual Firestore behavior or better: add an integration test against a test Firestore emulator to validate transactional behavior.
- Add tests to confirm likedBy is not present in public responses (privacy).
- Consider adding a test for pagination when getProductsLikedByUser returns many entries.

Other suggestions / small items
- Logging: add structured logs around like/unlike success/failure and include productId and userId (avoid logging full tokens).
- Monitoring: emit metrics for like/unlike counts and error rates.
- Rate-limiting: consider throttling like/unlike endpoints per user to reduce abuse/accidental repeated writes.
- Consistency script: provide an admin script to reconcile products' likes number with likedBy length and optionally fix mismatches. Document how to run it.
- API contract: ensure the OpenAPI schema for Product does not include likedBy (private) and document fields returned explicitly (likeCount, isLikedByUser).
- Backwards compatibility: confirm any clients relying on the old adjustLikes behavior are covered. You preserved adjustLikes; but add a short compatibility note in implementation doc about expected semantics.

Concrete code fixes (high level)
- Replace addLike() implementation with a transaction (read -> if not present then update likes and likedBy -> return snapshot). Do the same for removeLike in a transaction.
- Modify controllers to omit likedBy from responses.
- Add server-side limit/pagination for GET /api/products/user/liked.
- Add integration test (Firestore emulator) to assert idempotency under concurrent requests.

**Closing summary**

- The PR is a strong, well-documented feature addition with tests and Swagger updates. The major technical correctness/security issues to address before merging are:
  1) Fix addLike() idempotency bug (unconditional increment).
  2) Prevent leaking likedBy user IDs in API responses.
  3) Consider transactionizing removeLike() as well to avoid races.
  4) Add integration testing for concurrency; consider scaling & privacy implications and add pagination for GET user/liked.
- After addressing these points and adding a small reconciliation/monitoring plan, the feature should be safe to merge.

File: docs/LIKE_FEATURE_IMPLEMENTATION.md [hunk 1]
```diff
@@ -0,0 +1,339 @@
+# Like/Unlike Feature - User Product Preferences
```
Comment (yusuphjoluwasen) [inline docs/LIKE_FEATURE_IMPLEMENTATION.md]: code looks good thanks @1337Xcode, please can you remove this documentation and other code comments, thanks.

**Review by yusuphjoluwasen — CHANGES_REQUESTED** [review]
